src/components/icon/icon.js

```diff
@@ -20,25 +21,221 @@ angular.module('material.components.icon', [
+      $mdAria.expect(element, 'aria-label', ariaLabel);
+      $mdAria.expect(element, 'role', 'img');
+    }
+
+    if (attr.svgSrc) {
+      // Using External SVGs
+      $mdIcon(attr.svgSrc).then(function(svg) {
+        element.append(svg);
+      });
+    } else if (attr.svgIcon) {
+      // Use pre-configured named SVG.
+      $mdIcon(attr.svgIcon).then(function(svg) {
+        element.append(svg);
+      });
+    }
+  }
+}
+
+/**
+ * @ngdoc service
+ * @name $mdIconProvider
+ * @module material.components.icon
+ *
+ * @description
+ * `$mdIconProvider` is a factory that provides the $mdIcon service based on a
+ *  given configuruation.
+ *
+ * <hljs lang="js">
+ * var app = angular.module('app', ['ngMaterial']);
+ * app.config(function($mdIconProvider) {
+ *   $mdIconProvider.iconSet('social', 'my/app/social.svg')
+ *   $mdIconProvider.defaultIconSet('my/app/icons.svg')
+ *   $mdIconProvider.icon('android', 'my/app/android.svg')
+ *   $mdIconProvider.icon('work:chair', 'my/app/chair.svg')
+ * });
+ * </hljs>
+ */
+function MdIconProvider() {
+
+  var config = {
+    defaultIconSize: 24
+  };
+
+  this.iconSet = function(id, url, iconSize) {
+    config[id] = new Config(url, iconSize || config.defaultIconSize);
+    return this;
+  };
+
+  this.defaultIconSet = function(url, iconSize) {
+    var setName = '$default';
+    if (config[setName]) {
+      console.warn('Default icon set already defined');
+    } else {
+      config[setName] = new Config(url, iconSize || config.defaultIconSize);
+    }
+    return this;
+  };
+
+  this.icon = function(id, url, iconSize) {
+    id = id.match(/:/) ? id : '$default:' + id;
+    config[id] = new Config(url, iconSize || config.defaultIconSize);
+    return this;
+  };
+
+  this.defaultIconSize = function(iconSize) {
+    config.defaultIconSize = iconSize;
+  };
+
+  this.$get = ['$http', '$q', '$log', '$templateCache', function($http, $q, $log, $templateCache) {
+    return new MdIconService(config, $http, $q, $log, $templateCache);
+  }];
+
+  function Config(url, iconSize) {
+    this.url = url;
+    this.iconSize = iconSize;
+  };
+}
+
+
+/**
+ * @ngdoc service
+ * @name $mdIcon
+ * @module material.components.icon
+ *
+ * @description
+ * `$mdIcon` is a service used to retrieve SVG icons configured via $mdIconProvider
+ *  or a URL. The service returns a promise which resolves to a DOM element.
+ *
+ * <hljs lang="js">
+ * function SomeDirective($mdIcon) {
+ *   $mdIcon('android').then(function(android) {
+ *     element.append(android);
+ *   });
+ *   $mdIcon('work:chair').then(function(chair) {
+ *     element.append(chair);
+ *   });
+ * };
+ * </hljs>
+ */
+function MdIconService(config, $http, $q, $log, $templateCache) {
+
+  var iconCache = {};
+  var urlRegex = /[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
+  return function getIcon(id) {
+    id = id || '';
+
+    if (id.match(urlRegex)) {
```

Comment:
`urlRegex.test(id)` is more performant here because it doesn't actually separate the match into the various match groups. Not really something that needs to be changed, just pointing it out.